src/Microsoft.AspNet.OData.Shared/DefaultContainerBuilder.cs

```diff
@@ -83,15 +83,18 @@ public class DefaultContainerBuilder : IContainerBuilder
         /// <returns>The container built by this builder.</returns>
         public virtual IServiceProvider BuildContainer()
         {
-            /* "services.BuildServiceProvider()" returns IServiceProvider in Microsoft.Extensions.DependencyInjection 1.0 and ServiceProvider in Microsoft.Extensions.DependencyInjection 2.0
-            * (This is a breaking change)[https://github.com/aspnet/DependencyInjection/issues/550].
-            * To support both versions with the same code base in OData/WebAPI we decided to call that extension method using reflection.
-            * More info at https://github.com/OData/WebApi/pull/1082
-            */
-
-            MethodInfo buildServiceProviderMethod = typeof(ServiceCollectionContainerBuilderExtensions).GetMethod(nameof(ServiceCollectionContainerBuilderExtensions.BuildServiceProvider), new[] { typeof(IServiceCollection) });
+            try
```

Comment:
To avoid using exceptions as control flow logic and to avoid masking unexpected exceptions from `BuildServiceProvider`, we could do something like:
```
var buildServiceProviderMethod = typeof(ServiceCollectionContainerBuilderExtensions).GetMethod(
  nameof(ServiceCollectionContainerBuilderExtensions.BuildServiceProvider),
  new[] { typeof(IServiceCollection) });
if (buildServiceProviderMethod.ReturnType == typeof(ServiceProvider))
{
  return services.BuildServiceProvider();
}
else
{
  return (IServiceProvider)buildServiceProviderMethod.Invoke(null, new object[] { services });
}
```
To be even more helpful, we could do:
```
else if (buildServiceProviderMethod.ReturnType == typeof(IServiceProvider))
{
  return (IServiceProvider)buildServiceProviderMethod.Invoke(null, new object[] { services });
}
else
{
  throw new InvalidOperationException("No service provider method was found with the correct return type");
}
```

Reply:
alternatively, can we check for the conditions that would cause services.BuildServiceProvider() to fail, rather than waiting for it to fail and catching an exception?  what is the failure?

If we do have to catch, is there a specific exception that we can catch, rather than catching any exception?